src/libnnstreamer-edge/nnstreamer-edge-common.c

```diff
@@ -138,6 +138,276 @@ nns_edge_strdup_printf (const char *format, ...)
+
+  memset (meta, 0, sizeof (nns_edge_metadata_s));
+  return NNS_EDGE_ERROR_NONE;
+}
+
+/**
+ * @brief Internal function to free the list and items in metadata structure.
+ */
+int
+nns_edge_metadata_free (nns_edge_metadata_s * meta)
+{
+  nns_edge_metadata_node_s *node, *tmp;
+
+  if (!meta)
+    return NNS_EDGE_ERROR_INVALID_PARAMETER;
+
+  node = meta->list;
+  meta->list_len = 0;
+  meta->list = NULL;
+
+  while (node) {
+    tmp = node->next;
+
+    SAFE_FREE (node->key);
+    SAFE_FREE (node->value);
+    SAFE_FREE (node);
+
+    node = tmp;
+  }
+
+  return NNS_EDGE_ERROR_NONE;
+}
+
+/**
+ * @brief Internal function to set the metadata.
+ */
+int
+nns_edge_metadata_set (nns_edge_metadata_s * meta,
+    const char *key, const char *value)
+{
+  nns_edge_metadata_node_s *node;
+
+  if (!meta)
+    return NNS_EDGE_ERROR_INVALID_PARAMETER;
+
+  if (!STR_IS_VALID (key))
+    return NNS_EDGE_ERROR_INVALID_PARAMETER;
+
+  if (!STR_IS_VALID (value))
+    return NNS_EDGE_ERROR_INVALID_PARAMETER;
+
+  /* Replace old value if key exists in the list. */
+  node = nns_edge_metadata_find (meta, key);
+  if (node) {
+    char *val = nns_edge_strdup (value);
+
+    if (!val)
+      return NNS_EDGE_ERROR_OUT_OF_MEMORY;
+
+    SAFE_FREE (node->value);
+    node->value = val;
+    return NNS_EDGE_ERROR_NONE;
+  }
+
+  node = malloc (sizeof (nns_edge_metadata_node_s));
+  if (!node)
+    return NNS_EDGE_ERROR_OUT_OF_MEMORY;
+
+  memset (node, 0, sizeof (nns_edge_metadata_node_s));
+  node->key = nns_edge_strdup (key);
+  node->value = nns_edge_strdup (value);
+
+  if (!node->key || !node->value) {
+    SAFE_FREE (node->key);
+    SAFE_FREE (node->value);
+    SAFE_FREE (node);
+    return NNS_EDGE_ERROR_OUT_OF_MEMORY;
+  }
+
+  /* Prepend new node to the start of the list. */
+  meta->list_len++;
+  node->next = meta->list;
+  meta->list = node;
+
+  return NNS_EDGE_ERROR_NONE;
+}
+
+/**
+ * @brief Internal function to get the metadata in the list. Caller should release the returned value using free().
+ */
+int
+nns_edge_metadata_get (nns_edge_metadata_s * meta,
+    const char *key, char **value)
+{
+  nns_edge_metadata_node_s *node;
+
+  if (!value)
+    return NNS_EDGE_ERROR_INVALID_PARAMETER;
+
+  node = nns_edge_metadata_find (meta, key);
+  if (node) {
+    *value = nns_edge_strdup (node->value);
+    return (*value) ? NNS_EDGE_ERROR_NONE : NNS_EDGE_ERROR_OUT_OF_MEMORY;
+  }
+
+  return NNS_EDGE_ERROR_INVALID_PARAMETER;
+}
+
+/**
+ * @brief Internal function to copy the metadata.
+ */
+int
+nns_edge_metadata_copy (nns_edge_metadata_s * dest, nns_edge_metadata_s * src)
+{
+  nns_edge_metadata_node_s *node;
+  int ret;
+
+  if (!dest || !src)
+    return NNS_EDGE_ERROR_INVALID_PARAMETER;
+
+  nns_edge_metadata_free (dest);
+
+  node = src->list;
+  while (node) {
+    ret = nns_edge_metadata_set (dest, node->key, node->value);
+    if (ret != NNS_EDGE_ERROR_NONE) {
+      nns_edge_metadata_free (dest);
+      return ret;
```

Comment:
nns_edge_metadata_s tmp
tmp.list = dest->list;
loop: set(&tmp, ...)
if fail: return w/ error code (not touching dest)
if success: free dest, set dest->list = tmp.list

Reply:
updated